# Problem Description

There are three eletric power plants that supply to four cities such as New York, Boston, Philadelphia , NewJersy. Based on the data of the costs of sending 1 million kwh of electricity from a plant to a city, we want to find the optimal cost of meeting all citys’s peak power demand. Assuming their peak demand all occur at the same time and the transmission cost are assumed to be proportional to the distance the electricity need travel from plants to cities.


---
## Python Implementation

We import the Gurobi Python Module.

## OPTIMIZING ELECTRICITY TRANSMISSION COST

In [1]:
from gurobipy import *
#sets
plants=['Plant A','Plant B','Plant C' ]
cities =['NewYork','Boston','Philadelphia','NewJersy']
I = range(len(plants))
J = range(len(cities))

#data
Costs= [[8,12,7,3],
    [4,18,6,9],
    [14,2,10,15]]

Demand = [50,30,40,20]
Supply = [35,60,50]


# Transportation Problem  
m = Model('Trans Model')

X={}
for i in I:
    for j in J:
        X[i,j]=m.addVar()
        

m.setObjective(quicksum(Costs[i][j]*X[i,j] for i in I for j in J), GRB.MINIMIZE)

for j in J:
    m.addConstr(quicksum(X[i,j] for i in I) >= Demand[j])

for i in I:
    m.addConstr(quicksum(X[i,j] for j in J) <= Supply[i])

m.optimize()
print('Optimized Cost', m.objVal)

# Sensitivity Analysis
for j in J:
    print('-'*20)
    print("cities:",j,"Demand:",Demand[j])
    print("-"*20)
    for i in I:
        print("     plants", i, X[i,j].x)


Set parameter Username
Academic license - for non-commercial use only - expires 2022-11-17
Gurobi Optimizer version 9.5.0 build v9.5.0rc5 (win64)
Thread count: 2 physical cores, 4 logical processors, using up to 4 threads
Optimize a model with 7 rows, 12 columns and 24 nonzeros
Model fingerprint: 0x8b7903f0
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [2e+00, 2e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+01, 6e+01]
Presolve time: 0.01s
Presolved: 7 rows, 12 columns, 24 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.400000e+02   0.000000e+00      0s
       6    6.3500000e+02   0.000000e+00   0.000000e+00      0s

Solved in 6 iterations and 0.03 seconds (0.00 work units)
Optimal objective  6.350000000e+02
Optimized Cost 635.0
--------------------
cities: 0 Demand: 50
--------------------
     plants 0 0.0
     plants 1 50.0
     plants 2 0.0
--------------------
cities: 1 Demand: 30

|Plant| NewYork | Boston| Philadelphia| NewJersy| Supply (million kwh)|
|---| --- | --- | --- | ---| --- |
|A| 0  mKwh | 0 mKwh | 15 mKwh | 20 mKwh | 35 mKwh |
|B| 50 mKwh | 0 mKwh |  10 mKwh |  0 mKwh | 60 mKwh |
|C| 0 mKwh | 30 mKwh | 15 mKwh |  0 mKwh | 45 mKwh |
|Demand (million kwh)| 50 mKwh | 30 mKwh | 40 mKwh | 20 mKwh| 

The optimum cost of transmission is \$635.